### PR TITLE
Use enhanced MachineSpec to provision

### LINKED ISSUE #24: Feature: Add multiple nic support
Add support multiple nics on to the node machines.
This will help in setting the cluster in a way that the k8s API network and the data plane network could be possibly isolated

### LINKED ISSUE #28: Feature:  support static IP in machine definition
There are times when there is a desire to create machine nodes with static IPs.

### PR DESCRIPTION
**Pull request by sidharthsurana**
Start using the new `MachineSpec` definition for creating the instance on vsphere

The code now acts accordingly based on the availability of the new cloud-init datasource
`VMwareGuestInfo` inside the template used. This availability can be set by the users in
the spec using the flag `vsphereCloudInit` set to `true`. For usual upstream version of
ubuntu cloud image that has been used so far, this flag need not be set at all and the
controller will utilize the OVF properties available on the ubuntu cloud image to inject
the cloud-init data.
**Note**: The new static IP support is only available with the templates that have the vsphere
cloud-init datasource. DHCP based IPs will continue to work as they have in the past.

Going forward we prefer users creating a custom ubuntu cloud image that also has this new
cloud-init datasource pre-installed. As that will give them the ability to use static IPs
as well if they want.

See https://github.com/akutz/cloud-init-vmware-guestinfo for more info on the new cloud-init
datasource

* Adds support for multiple nics
* Adds support for static IP

Resolves #24
Resolves #28

Change-Id: Id2dee12b382d57b6591f840dadd77a1c287667c4

### CODE REVIEW

**Comment by frapposelli** [comment]
Please rebase the PR 🙏

File: pkg/cloud/vsphere/provisioner/govmomi/create.go [hunk 1]
```diff
@@ -147,112 +151,158 @@ func (pv *Provisioner) cloneVirtualMachine(s *SessionContext, cluster *clusterv1
...
-			Info: &types.VAppPropertyInfo{
-				Key:   p.Key,
-				Id:    p.Id,
-				Value: defaultValue,
-			},
-		}
-		if p.Id == "user-data" {
-			prop.Info.Value = userData
+
+	if machineConfig.MachineSpec.VsphereCloudInit {
```
[the remark this answers is not in the document]
Comment (sidharthsurana) [reply pkg/cloud/vsphere/provisioner/govmomi/create.go]: yes, the user can specify this flag in the spec if the template they are using has the cloud-init datasource pre-installed

File: pkg/cloud/vsphere/provisioner/govmomi/create.go [hunk 2]
```diff
@@ -147,112 +151,158 @@ func (pv *Provisioner) cloneVirtualMachine(s *SessionContext, cluster *clusterv1
...
+	diskChange := false
 	for _, dev := range disks {
 		disk := dev.(*types.VirtualDisk)
-		if disk.DeviceInfo.GetDescription().Label == disklabel {
-			newsize, err := strconv.ParseInt(machineConfig.MachineVariables["disk_size"], 10, 64)
-			if err != nil {
-				return err
+		if newSize, ok := diskMap[disk.DeviceInfo.GetDescription().Label]; ok {
+			if disk.CapacityInBytes > vsphereutils.GiBToByte(newSize) {
+				return errors.New("[FATAL] Disk size provided should be more than actual disk size of the template. Please correct the machineSpec to proceed")
```
[the remark this answers is not in the document]
Comment (sidharthsurana) [reply pkg/cloud/vsphere/provisioner/govmomi/create.go]: This check is essentially to verify that if the user has specified a disk size that needs to be strictly bigger than the existing disk in the template

File: pkg/cloud/vsphere/provisioner/govmomi/create.go [hunk 3]
```diff
@@ -147,112 +151,158 @@ func (pv *Provisioner) cloneVirtualMachine(s *SessionContext, cluster *clusterv1
...
+			disk.CapacityInBytes = vsphereutils.GiBToByte(newSize)
+			diskspec := &types.VirtualDeviceConfigSpec{}
+			diskspec.Operation = types.VirtualDeviceConfigSpecOperationEdit
+			diskspec.Device = disk
+			deviceSpecs = append(deviceSpecs, diskspec)
 		}
 	}
-	if targetdisk == nil {
-		return fmt.Errorf("Could not locate the disk with label %s", disklabel)
+	if !diskChange && len(machineConfig.MachineSpec.Disks) > 0 {
```
[the remark this answers is not in the document]
Comment (sidharthsurana) [reply pkg/cloud/vsphere/provisioner/govmomi/create.go]: the same reason as above.

**Comment by sflxn** [comment]
lgtm

**Pull request merged**